File: Python_Algorithm/Baek/9663.py
N = int(input())
arr = [[0 for i in range(N)] for j in range(N)]
result = 0
def check(x,y):
    #col check
    for i in range(x):
        if arr[i][y] == 1:
            return False
    # left cross check
    x_,y_ = x,y
    while 0 <= x_ <N and 0<=y_<N:
        if arr[x_][y_] == 1:
            return False
        x_ -= 1
        y_ -= 1
    # right cross check
    x_,y_ = x,y
    while 0 <= x_ <N and 0<=y_<N:
        if arr[x_][y_] == 1:
            return False
        x_ -= 1
        y_ += 1
    return True

def dfs(x):
    global result
    if x == N:
        result += 1
        return
    for i in range(N):
        if check(x,i):
            arr[x][i] = 1
            dfs(x+1)
            arr[x][i] = 0
dfs(0)
print(result)

File: C++_Algorithm/Baek/9663.cpp
#include <iostream>

using namespace std;

int arr[15][15];
int result = 0;
int N;
bool check(int x, int y) {
	for (int i = 0; i < x; i++) {
		if (arr[i][y] == 1)
			return false;
	}
	int xx = x;
	int yy = y;
	while (xx >= 0 && yy >= 0) {
		if (arr[xx][yy] == 1)
			return false;
		xx--;
		yy--;
	}

	xx = x;
	yy = y;
	while (xx >= 0 && yy < N) {
		if (arr[xx][yy] == 1)
			return false;
		xx--;
		yy++;
	}
	return true;
}

void dfs(int x) {
	if (x == N) {
		result++;
		return;
	}
	for (int i = 0; i < N; i++) {
		if (check(x, i)) {
			arr[x][i] = 1;
			dfs(x + 1);
			arr[x][i] = 0;
		}
	}
}
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	cin >> N;
	dfs(0);
	cout << result << "\n";


	return 0;
}